src/dataflow/src/render/join.rs

```diff
@@ -294,4 +281,62 @@ where
             panic!("render_join called on invalid expression.")
         }
     }
+
+    /// Joins the next input to the arranged version of the join of previous inputs.
+    /// This is split into its own method to enable reuse of code with different
+    /// types of `prev_keyed`.
+    fn differential_join<J>(
+        &mut self,
+        prev_keyed: J,
+        next_input: &RelationExpr,
+        next_keys: &[ScalarExpr],
+        next_vals: Vec<usize>,
+    ) -> (Collection<G, Row>, Collection<G, DataflowError>)
```

Comment:
This looks great! Without suggesting that we *should* change it, what do you think about the idea of making the method generic over the type of *other* arrangement, so that it can be called with `self.arrangement(next_input, next_keys)`? My main motivation is keeping the assert in the method closer to where it is enforced, but also it would draw down the size of the body of the method. If we believe the `TODO`, we'll eventually want to jam some more logic in here (perhaps a `MapFilterProject` sort of thing) and probably having one version of the join body ends up being the right answer.

This can totally wait, but I did want to get your thoughts!

Reply:
Just making `differential_join` generic over the type of the other arrangement results in code duplication where every type of `prev_keyed` has to account for the two type of `next_input` there are. But I think I have made a happy compromise by having

```
fn differential_join_inner<J, Tr2> (prev_key, next_input_arranged, next_vals) {
// generic over both arrangements
// only one version of join body
}

fn differential_join<J>(prev_keyed, next_input, next_keys, next_vals) {
  //enforce assert
  // look up self.arrangement(next_input, next_keys) and run differential_join_inner
}
```